test/extensions/filters/http/ext_proc/ext_proc_integration_test.cc

```diff
@@ -699,6 +699,68 @@ TEST_P(ExtProcIntegrationTest, GetAndSetOnlyTrailersOnResponse) {
   EXPECT_THAT(*(response->trailers()), SingleHeaderValueIs("x-modified-trailers", "xxx"));
 }
 
+// Test the filter with a response body callback enabled using an
+// an ext_proc server is okay to buffer more than buffer limit on last onData(xxx, end_stream=true)
+// call, while waiting for the headers response.
+TEST_P(ExtProcIntegrationTest, ResponseBodySentWithEndStreamBiggerThanBufferLimitWillBeBuffered) {
+  // Make sure that we have control over when buffers will fill up.
+  config_helper_.setBufferLimits(/*upstream_buffer_limit=*/10, /*downstream_buffer_limit=*/100);
+  proto_config_.mutable_processing_mode()->set_response_body_mode(ProcessingMode::BUFFERED);
+  initializeConfig();
+  HttpIntegrationTest::initialize();
+  auto response = sendDownstreamRequest(absl::nullopt);
+  processRequestHeadersMessage(*grpc_upstreams_[0], true, absl::nullopt);
+
+  ASSERT_TRUE(fake_upstreams_[0]->waitForHttpConnection(*dispatcher_, fake_upstream_connection_));
+  ASSERT_TRUE(fake_upstream_connection_->waitForNewStream(*dispatcher_, upstream_request_));
+  ASSERT_TRUE(upstream_request_->waitForEndStream(*dispatcher_));
+  upstream_request_->encodeHeaders(Http::TestResponseHeaderMapImpl{{":status", "200"}}, false);
+
+  // Make sure onData(end_stream=true) is called while waiting for header response.
+  absl::Notification body_sent_notification;
+  absl::Notification headers_sent_notification;
+
+  // Pretend to be the upstream.
+  Thread::ThreadPtr backend_thread = api_->threadFactory().createThread([&, this]() {
+    // Make sure that the headers is sent to server.
+    headers_sent_notification.WaitForNotification();
+    // Send body in one big chunk.
+    upstream_request_->encodeData(10000, true);
+    body_sent_notification.Notify();
+  });
+
+  processResponseHeadersMessage(
+      *grpc_upstreams_[0], false, [&](const HttpHeaders&, HeadersResponse& headers_resp) {
+        headers_sent_notification.Notify();
+        auto* content_length =
+            headers_resp.mutable_response()->mutable_header_mutation()->add_set_headers();
+        content_length->mutable_header()->set_key("content-length");
+        content_length->mutable_header()->set_value("13");
+        // Make sure encodeData is called with end_stream before header response is sent.
```

Comment:
Are you sure this is guaranteed? It looks like this races with the Envoy's worker thread. It may either process the data sent by the backend_thread first, or the response from the ext_proc processing response headers.

It looks like you may need to wait for the data to arrive to ext_proc before completing the callback for the headers processing.

Reply:
IIUC, the encodeData call at line 728 already pushed the data in one shot to ext_proc filter?
     upstream_request_->encodeData(10000, true);

This notification is notified after the mentioned finishes, so I can assume the data is pushed to the filter already.

Reply:
I think @yanavlasov 's point might be that `upstream_request_->encodeData()` simulates the upstream sending response data but doesn't guarantee that ext_proc filter in the Envoy's worker thread has actually been called and received that data.

But in order to guarantee that reliably, it seems like we'd need an internal sync point in ext_proc or HCM, which seems tricky since this level of test does not have direct access to the ext_proc filter. @yanavlasov do you know offhand of other integration tests that orchestrate timing like this, and what hooks they use?

Unfortunately we can't rely on "client receiving body bytes" as a causal timing dependency to ensure that they have passed through the filter chain, since we're buffering the bytes...

Reply:
ah, I see it now. 
this is integration test. Bytes get into Envoy filter chain through network connections. 
Let me see what kinda of sync mechanism is available in integration test, besides "sleep for a second".

Reply:
`upstream_request_->encodeData(10000, true);` will schedule a callback into the upstream worker thread, that will write the data into the socket, and then Envoy worker thread will need to get event and read that data.

Reply:
yeah, but depends on the network layer, it may or maynot end up with a single encodeData(, true) call on the router, as well as the ext_proc filter.